Fix - Add support for multiple AssemblyLoadContexts

derstpal opened the pull request:
Hello,

We are using a Plugin pattern for our dotnet application. Each plugin can use its own GRPC contract in isolation. The contract is loaded inside an AssemblyLoadContext. Plugins also share some common dependencies like protobuf-net.Grpc

When we Create a proxy on a same contract but loaded from different ALCs, we have various Reflection Type exceptions.
This PR contains a fix for this case and remain compatible on one default ALC and also other targetted frameworks.

Some docs:
[Contextual Reflection](https://github.com/dotnet/coreclr/blob/v3.0.0/Documentation/design-docs/AssemblyLoadContext.ContextualReflection.md)

Commit:
Fix - Scope Reflection into the ALC owning the Proxy. One single DynamicModule per ALC.

Commit:
Fix - Scope Reflection into the ALC owning the Proxy. One single DynamicModule per ALC.

Commit:
Add unit tests

Commit:
Add unit tests

Commit:
more precise fix

Commit:
more precise fix

mgravell committed:
Rebase fixups: keep the test proxy out of the ship list, net8+net10

Mechanical only; the fix itself is unchanged.

* protobuf-net.Grpc.Tests.TestProxy moves from src/ to tests/. src/** *is* the
  ship list - Build.csproj uses exactly that glob when Packing=true, and nothing
  opts out - so a test fixture there would have published as a seventh NuGet
  package. Also marked IsPackable=false, belt and braces. Verified: pack still
  produces exactly the same six packages.

* Retargeted net6.0;net8.0 -> net8.0;net10.0, matching the suite that consumes it
  (net472;net8.0;net10.0, referencing this only off net4x); net6.0 is out of
  support and was unlike anything else here.

* The solution entry goes in protobuf-net.Grpc.slnx; the rebase had landed the
  original .sln edit on protobuf-net.Grpc.sln.old, which is the stale leftover.

* The test csproj conflict was churn only - the PR predates the move to
  net472/net8.0/net10.0 and xunit.v3 - so main's version stands, plus the one
  genuine addition: the TestProxy project reference.

Full solution builds; 211/211 on net8.0 and net10.0, 206/206 on net472.

The three AssemblyLoadContext tests are known to be able to fail: reverting
ProxyEmitter/ProxyModuleHelper alone gives

    Object of type 'Grpc.Core.Marshaller`1[...TestProxy.In]' cannot be
    converted to type 'Grpc.Core.Marshaller`1[...TestProxy.In]'

i.e. the same type from two load contexts - the symptom the PR set out to fix.

mgravell commented:
Thanks for this, and apologies for the long silence — the scenario is legitimate and the PR is a real fix.

I've rebased the branch onto current `main` and added one fixup commit, so it's ready to review. **Note this was a force-push** (a rebase rewrites SHAs); your three commits are unchanged in content and still attributed to you, they just have new hashes.

### Why the rebase was needed

The PR predates a few things: the solution moved to `.slnx`, and the test project moved to `net472;net8.0;net10.0` on xunit.v3. The only conflict was that churn — main's version stands, plus your one genuine addition (the `TestProxy` project reference).

### The fixup commit

- **`protobuf-net.Grpc.Tests.TestProxy` moved `src/` → `tests/`**, and marked `IsPackable=false`. This is the one that mattered: `src/**` *is* the ship list — `Build.csproj` uses exactly that glob when `Packing=true` — so a test fixture there would have been published to nuget.org as a seventh package. Verified `pack` still produces exactly the same six.
- **Retargeted `net6.0;net8.0` → `net8.0;net10.0`**, matching the suite that consumes it; net6.0 is out of support.
- **Solution entry into `.slnx`** (the rebase had landed the original `.sln` edit on `protobuf-net.Grpc.sln.old`, which is a stale leftover).
- **A comment** on `_proxyModules` explaining why strong ALC keys are safe — a collectible ALC only unloads on an explicit `Unload()`, and the `Unloading` handler removes it before collection. It reads like a leak at a glance and deserved a line saying why it isn't.

### On the fix itself

I checked the part I couldn't take on faith — whether `EnterContextualReflection()` really governs which ALC a *dynamically defined* assembly lands in, as opposed to just `Assembly.Load` resolution. Your tests answer it: reverting only `ProxyEmitter`/`ProxyModuleHelper` and leaving everything else gives

```
Object of type 'Grpc.Core.Marshaller`1[...TestProxy.In]' cannot be converted to
type 'Grpc.Core.Marshaller`1[...TestProxy.In]'
```

— the same type from two load contexts, which is exactly the symptom. Tests that demonstrably fail without the change are the most valuable thing in a PR like this, so thank you for including them.

Full solution builds; 211/211 on net8.0 and net10.0, 206/206 on net472.

mgravell committed:
Use #if NET rather than #if NET6_0_OR_GREATER

Scoped to the four gates this PR introduces; the rest of the repo still uses
assorted version floors, and normalising those is a separate change rather than
something to smuggle in here.

Equivalent for this project - verified per-TFM rather than assumed, NET,
NET6_0_OR_GREATER and NETCOREAPP3_0_OR_GREATER are all defined on exactly net8.0
and net10.0, and on none of net472, netstandard2.0 or netstandard2.1 - but it
says the intended thing: "modern .NET, as opposed to netfx or netstandard".
NET6_0_OR_GREATER implies a version floor that is not real; nothing here needs
net6 specifically, and the gate reads as though net5.0 were excluded on purpose
when it simply is not in the TFM list. It also cannot go stale if a TFM is added.

(The strictly accurate floor would be NETCOREAPP3_0_OR_GREATER, since
AssemblyLoadContext.EnterContextualReflection arrived in .NET Core 3.0; that
buys nothing given netcoreapp3.x will not be coming back.)

211/211 net8.0 and net10.0, 206/206 net472.

mgravell commented:
One more commit: the four conditional-compilation gates this PR introduces now use `#if NET` rather than `#if NET6_0_OR_GREATER`.

Equivalent here — verified per-TFM rather than assumed; `NET`, `NET6_0_OR_GREATER` and `NETCOREAPP3_0_OR_GREATER` are all defined on exactly `net8.0` and `net10.0`, and on none of `net472`, `netstandard2.0` or `netstandard2.1`. But `#if NET` says the intended thing — *modern .NET, as opposed to netfx or netstandard* — where `NET6_0_OR_GREATER` implies a version floor that isn't real, and reads as though net5.0 were excluded deliberately when it just isn't in the TFM list. It also can't go stale if a TFM is added.

Deliberately scoped to this PR only; the rest of the repo still uses assorted version floors, and normalising those is a follow-up rather than something to fold in here.

Still 211/211 on net8.0 and net10.0, 206/206 on net472.

**Follow-up for after this merges** (noting it here so it isn't lost): sweep the remaining version gates repo-wide — `src` currently has `NETCOREAPP3_1_OR_GREATER` ×3, `NET5_0_OR_GREATER` ×3 and `NET6_0_OR_GREATER` ×4, several of which are the same "modern .NET" intent spelled three different ways.

mgravell committed:
Fix CI: the fixture project is not a test project

Directory.Build.props infers IsTestProject from the project *name*:

    <IsTestProject>$(MSBuildProjectName.Contains('Test'))</IsTestProject>

protobuf-net.Grpc.Tests.TestProxy contains "Test", so `dotnet test` on the
traversal tried to run VSTest against a fixture library with no test adapter and
aborted the run. Setting IsTestProject=false explicitly overrides the heuristic.

Not introduced by moving the project to tests/ - the name has always contained
"Test", and the traversal included src/** unconditionally, so this would have
failed identically in the original layout. It was simply never run.

Verified with the exact commands the workflow uses rather than an approximation,
which is how it slipped through the first time: I checked Debug and grepped for
"Failed!", and the symptom is "Test Run Aborted" plus "(VSTest target(s)) --
FAILED" in Release.

    dotnet build Build.csproj --no-restore -c Release /p:CI=true /p:GeneratePackageOnBuild=false
    dotnet test Build.csproj --no-build -c Release /p:CI=true --verbosity normal
      -> 12 "Test Run Successful.", zero FAILED

Debug is green too, and pack still produces exactly six packages.
